File: 0414-third-maximum-number/0414-third-maximum-number.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var thirdMax = function(nums) {
    let max = -Infinity, smax = -Infinity, tmax = -Infinity;

    for (let num of nums) {
        if (num === max || num === smax || num === tmax) continue;

        if (num > max) {
            tmax = smax;
            smax = max;
            max = num;
        } else if (num > smax) {
            tmax = smax;
            smax = num;
        } else if (num > tmax) {
            tmax = num;
        }
    }

    return tmax === -Infinity ? max : tmax;
};

File: 0414-third-maximum-number/0414-third-maximum-number.cpp
class Solution {
public:
    int thirdMax(vector<int>& nums) {
        set<int> s(nums.begin(), nums.end());  
        // int first = *v.begin()+3;
        if(s.size()<3){
            return *s.rbegin();
        }
        auto it = s.rbegin();
        for(int i=0;i<2;i++){
            it++;
        }
        return *it;                   
    }
};
